haskell/set.bzl

```diff
@@ -0,0 +1,111 @@
+"""Immutable sets that support efficient merging, traversal, and membership
+check.
+"""
+
+def _new():
+  """Create an empty set.
+  """
+  return struct(_set_items = dict())
+
+def _is_member(s, e):
+  """Return true if `e` is in the set `s`.
+
+  Args:
+    s: The set to inspect.
+    e: The element to search for.
+
+  Result:
+    Bool, true if `e` is in `s`, false otherwise.
+  """
+  e in s._set_items.get(e)
+
+def _insert(s, e):
+  """Insert an element into the set.
+
+  Args:
+    s: Set to insert new element into.
+    e: The element to insert.
+
+  Result:
+    A copy of set `s` with `s` element added.
+  """
+  r = dict(s._set_items)
+  r[e] = None
+  return struct(_set_items = r)
```

Comment:
I think `return struct(_set_items = dict(s._set_items, e = None))` would work if we want to be more concise without explicit binding to `r` and mutation.